### PR TITLE
[WIP][CALCITE-2972] Upgrade jetty to 9.4.15.v20190215

### PR DESCRIPTION
**Pull request by risdenk**
Upgrade jetty to 9.4.15.v20190215 and remove deprecations associated with the upgrade.

### CODE REVIEW

**Comment by risdenk** [comment]
Not ready for review yet but putting in place to push changes as I make them.

**Comment by risdenk** [comment]
So I'm definitely getting by butt kicked on this. I am going to push my changes that are definitely broken. 

https://github.com/eclipse/jetty.project/blob/jetty-9.4.x/jetty-client/src/test/java/org/eclipse/jetty/client/util/SPNEGOAuthenticationTest.java#L141

File: server/src/main/java/org/apache/calcite/avatica/server/HttpServer.java [hunk 1]
```diff
@@ -337,17 +343,23 @@ protected ConstraintSecurityHandler configureSpnego(Server server,
...
     final String principal = Objects.requireNonNull(config.getKerberosPrincipal());
-
-    // A customization of SpnegoLoginService to explicitly set the server's principal, otherwise
-    // we would have to require a custom file to set the server's principal.
-    PropertyBasedSpnegoLoginService spnegoLoginService =
-        new PropertyBasedSpnegoLoginService(realm, principal);
+    final String principalWithoutRealm = principal.split("@", 2)[0];
+    final String[] principalParts = principalWithoutRealm.split("/", 2);
+    final String serviceName = principalParts[0];
+    final String hostname = principalParts[1];
```
Comment (risdenk) [inline server/src/main/java/org/apache/calcite/avatica/server/HttpServer.java]: Currently Avatica server is configured with a principal. The Jetty implementation builds a principal from the service name (ie: `avatica`) and the hostname (ie: `localhost`). So breaking apart here to get into necessary parts.

File: server/src/main/java/org/apache/calcite/avatica/server/HttpServer.java [hunk 2]
```diff
@@ -337,17 +343,23 @@ protected ConstraintSecurityHandler configureSpnego(Server server,
...
+    final Path keytabPath = config.getKerberosKeytab();
+
+    ConfigurableSpnegoLoginService spnegoLoginService =
+        new ConfigurableSpnegoLoginService(realm, new TestAuthorizationService());
```
Comment (risdenk) [inline server/src/main/java/org/apache/calcite/avatica/server/HttpServer.java]: `ConfigurableSpnegoLoginService` takes an `AuthorizationService` to configure roles. Not sure why this is necessary at all. This is one of the parts I am really confused on. I tried with `HashLoginService` but that didn't work OOTB.

Here is an example test:
https://github.com/eclipse/jetty.project/blob/jetty-9.4.x/jetty-client/src/test/java/org/eclipse/jetty/client/util/SPNEGOAuthenticationTest.java#L141

File: server/src/main/java/org/apache/calcite/avatica/server/HttpServer.java [hunk 3]
```diff
@@ -337,17 +343,23 @@ protected ConstraintSecurityHandler configureSpnego(Server server,
...
+    spnegoLoginService.setServiceName(serviceName);
+    spnegoLoginService.setHostName(hostname);
+    spnegoLoginService.setKeyTabPath(keytabPath);
```
Comment (risdenk) [inline server/src/main/java/org/apache/calcite/avatica/server/HttpServer.java]: It looks like Avatica server currently supports logging in without a keytab. Jetty wants to now manage the Kerberos lifecycle and login.

File: server/src/main/java/org/apache/calcite/avatica/server/HttpServer.java [hunk 4]
```diff
@@ -337,17 +343,23 @@ protected ConstraintSecurityHandler configureSpnego(Server server,
...
 
     // Roles are "realms" for Kerberos/SPNEGO
     final String[] allowedRealms = getAllowedRealms(realm, config);
 
     return configureCommonAuthentication(Constraint.__SPNEGO_AUTH,
-        allowedRealms, new AvaticaSpnegoAuthenticator(), realm, spnegoLoginService);
+        allowedRealms, new ConfigurableSpnegoAuthenticator(), realm, spnegoLoginService);
```
Comment (risdenk) [inline server/src/main/java/org/apache/calcite/avatica/server/HttpServer.java]: Fun fact, can't move to `ConfigurableSpnegoAuthenticator` without ALSO moving to `ConfigurableSpnegoLoginService`. So these are tied together. I tried to separate out the changes but get a class cast exception.

**Comment by risdenk** [comment]
Since the existing implementation will work (not removed from Jetty yet just deprecated), I am thinking to just suppress the warnings and do the upgrade. I can open a separate PR for that.

File: server/src/main/java/org/apache/calcite/avatica/server/HttpServer.java [hunk 5]
```diff
@@ -784,13 +795,6 @@ public HttpServer build() {
         if (usingTLS) {
           throw new IllegalArgumentException("TLS has not been tested wtih SPNEGO");
         }
-        if (null != keytab) {
```
Comment (risdenk) [inline server/src/main/java/org/apache/calcite/avatica/server/HttpServer.java]: Jetty handles the login via Kerberos now.

File: server/src/main/java/org/apache/calcite/avatica/server/HttpServer.java [hunk 6]
```diff
@@ -950,25 +963,6 @@ private AvaticaServerConfiguration buildUserAuthenticationConfiguration(Builder
         }
       };
     }
-
-    private Subject loginViaKerberos(Builder b) {
```
Comment (risdenk) [inline server/src/main/java/org/apache/calcite/avatica/server/HttpServer.java]: Jetty handles the login via Kerberos now.

**Comment by risdenk** [comment]
I created https://github.com/apache/calcite-avatica/pull/93 for the suppression route.

File: server/src/main/java/org/apache/calcite/avatica/server/HttpServer.java (hunk 3 above)
Comment (joshelser) [reply server/src/main/java/org/apache/calcite/avatica/server/HttpServer.java]: Ugh. That's going to make unit testing harder.

**Comment by risdenk** [comment]
Closing this since didn't go this route for CALCITE-2972.

**Pull request closed without merging**